examples/xmllib/main.py

```diff
@@ -0,0 +1,77 @@
+# noqa: INP001 (ignore add __init__.py)
+
+
+from pathlib import Path
+
+import pandas as pd
+
+from dsp_tools import xmllib
+
+data_folder = Path("examples/xmllib/data")
+
+
+def _make_books(res_collection: xmllib.ResourceCollection) -> None:
+    df = pd.read_csv(data_folder / "Book.csv")
+    for _, row in df.iterrows():
+        one_book = _make_one_book(row)
+        res_collection.with_Resource(one_book)
+
+
+def _make_one_book(row: pd.Series) -> xmllib.Resource:
+    res = xmllib.Resource(res_id=row["id"], restype=":Book", label=row["label"])
+    titles = row["hasTitle"].split("||")
+    titles = [x.strip() for x in titles]
+    res.with_SeveralSimpleTexts(
+        prop_name=":hasTitle",
+        values=titles,
+    ).with_IntValue(
+        prop_name=":hasNumberOfPages",
+        value=row["hasNumberOfPages"],
+    ).with_SimpleText(prop_name=":hasAuthor", value=row["hasAuthor"])
+    return res
+
+
+def _make_chapters(res_collection: xmllib.ResourceCollection) -> None:
+    df = pd.read_csv(data_folder / "Chapter.csv")
+    for _, row in df.iterrows():
+        one_chapter = _make_one_chapter(row)
+        res_collection.with_Resource(one_chapter)
+
+
+def _make_one_chapter(row: pd.Series) -> xmllib.Resource:
+    return (
+        xmllib.Resource(
+            res_id=row["id"],
+            restype=":Chapter",
+            label=row["label"],
+        )
+        .with_SimpleText(
+            prop_name=":hasTitle",
+            value=row["hasTitle"],
+        )
+        .with_IntValue(
+            prop_name=":hasChapterNumber",
+            value=row["hasChapterNumber"],
+        )
+        .with_LinkValue(prop_name=":isPartOfBook", value=row["isPartOfBook"])
+    )
+
+
+def _add_images(res_collection: xmllib.ResourceCollection) -> None:
+    files = [x for x in data_folder.glob("*.jpg")]
+    for f in files:
+        res_collection.add_FileValueToResource(res_id=f.stem, filepath=str(f))
+
+
+def main() -> None:
+    """Creates an XML file from the csv files in the data folder."""
+    resources = xmllib.ResourceCollection()
+    _make_books(resources)
+    _make_chapters(resources)
+    _add_images(resources)
+    root = xmllib.XMLRoot(shortcode="0001", default_ontology="onto", resource_collection=resources)
+    root.write_file("examples/xmllib/data.xml")
```

Comment:
this still doesn't use the fluent interface pattern. I would say it should be like this:

```suggestion
    xml = xmllib.XMLRoot(shortcode="0001", default_ontology="onto")
        .add(_make_books())
        .add(_make_chapters())
        .add(_make_images())
    xml.write_file("examples/xmllib/data.xml") 
```

then you could even do 

```python
    xmllib.XMLRoot(shortcode="0001", default_ontology="onto")
        .add(_make_books())
        .add(_make_chapters())
        .add(_make_images())
        .write_file("examples/xmllib/data.xml") 
```

Reply:
but in terms of architecture (and testability) it would be better to have it like this:

```python
def main() -> None:
    books_data = _load_books_data()
    chapters_data = ...
    images_data = ...
    xmllib.XMLRoot(shortcode="0001", default_ontology="onto")
        .add(_make_books(books_data))
        .add(_make_chapters(chapters_data))
        .add(_make_images(images_data))
        .write_file("examples/xmllib/data.xml") 
```

Reply:
what do you imagine `books_data` to look like?

Reply:
I see that you would like the fluent interface pattern everywhere. I am concerned that RDU won't like that. For example some prefer to try out only one class, or they want a separate file for each class (because it is easier to read when the file is super big). So they have to be able to add data without being forced to chain (I just noticed that I don't provide that optionality, but that is a detail).

Reply:
There is never an urge to chain. AFAIU, you are always free to either chain or not, that's the big advantage. Because the object is modified in place, plus it is returned. So either you rely on the returning, by chaining your functions, or you rely on in-place-modification, by NOT chaining.

Reply:
I am very sceptical about in-place modification. Normally I would be 100% against it, but I do see that this may lead to errors when people try to do that (though the pattern shouldn't be unfamiliar: `str.replace()` is also not modifying in place).

But I very much agree that it the implementation has to be done in such a way that no one is forced to do chaining.  
But in order to convey the concept of the fluent interface pattern - and from my understanding, that's what this PR is about - this pattern should be used as consistently and ubiquitously as possible, to make it clear what it is about.

Reply:
> what do you imagine `books_data` to look like?

that would be the dataframe loaded from the CSV.  
maybe something like:

```python
books_data = _load_data("books.csv")
```

the point of that being that `_make_books()` then becomes a pure function that has no side effects and doesn't depend on the file system, which makes it a breeze to test.